### PR TITLE
Use embedded previews of ZIP-like files

### PR DESCRIPTION
**Pull request by juliusknorr**
Some filetypes like OpenDocument are basically just zip files and contain a preview image already that we can extract to provide previews without requiring external dependencies.

This implements an abstract class to extract as well as two providers for OpenDocument and Krita files.

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]
That is so awesome!!
Code looks good!

Maybe some tests? :see_no_evil:

**Comment by juliushaertl** [comment]
Hm, not sure how I can properly write tests for that as the preview providers don't support DI yet.

**Comment by skjnldsv** [comment]
> Hm, not sure how I can properly write tests for that as the preview providers don't support DI yet.

Ah ok :)

**Comment by skjnldsv** [comment]
> Code looks good!

Ah, that was not done! I was happy this looked so simple with the first commit :see_no_evil:

**Comment by juliushaertl** [comment]
It is now, I just forgot to push a new file and bump the autoloader.

File: lib/private/Preview/Krita.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (nickvergessen) [inline lib/private/Preview/Krita.php:3]: Nope

**Comment by kesselb** [comment]
Nice :+1: Works for me :tada: 

https://github.com/nextcloud/server/blob/53db05a1f67fc974dba904ec158b2d67fa72df95/config/config.sample.php#L981-L1032 Please update accordingly.

File: lib/private/Preview/Bundled.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * Extracts a preview from files that embed them in an ZIP archive
+ */
+abstract class Bundled extends ProviderV2 {
+
+	protected function extractThumbnail(File $file, $path): ?IImage {
+		$sourceTmp = \OC::$server->getTempManager()->getTemporaryFile();
+		$targetTmp = \OC::$server->getTempManager()->getTemporaryFile();
+
+		try {
+			$content = $file->fopen('r');
```
Comment (kesselb) [inline lib/private/Preview/Bundled.php:40]: `File` has a `getContent` method returning a string with the content.

File: lib/private/Preview/Bundled.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			$content = stream_get_contents($content);
```
Comment (kesselb) [inline lib/private/Preview/Bundled.php:41]: If you keep `stream_get_contents` please check for false and return null.

**Review by kesselb — APPROVED** [review]

**Comment by rullzer** [comment]
@juliushaertl please address the comments and we can get this in

**Comment by juliushaertl** [comment]
Done

**Pull request merged**